x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionFieldWildcardQuery.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.versionfield;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.AutomatonQuery;
+import org.apache.lucene.search.WildcardQuery;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.automaton.Automata;
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.Operations;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A variation of the {@link WildcardQuery} than skips over meta characters introduced using {@link VersionEncoder}.
+ */
+class VersionFieldWildcardQuery extends AutomatonQuery {
+
+    private static final byte WILDCARD_STRING = '*';
+
+    private static final byte WILDCARD_CHAR = '?';
+
+    VersionFieldWildcardQuery(Term term) {
+        super(term, toAutomaton(term), Integer.MAX_VALUE, true);
+    }
+
+    private static Automaton toAutomaton(Term wildcardquery) {
+        List<Automaton> automata = new ArrayList<>();
+
+        BytesRef wildcardText = wildcardquery.bytes();
+        boolean containsPreReleaseSeparator = false;
+
+        for (int i = 0; i < wildcardText.length;) {
+            final byte c = wildcardText.bytes[wildcardText.offset + i];
+            int length = Character.charCount(c);
+
+            switch (c) {
+                case WILDCARD_STRING:
+                    automata.add(Automata.makeAnyString());
+                    break;
+                case WILDCARD_CHAR:
+                    // this should also match leading digits, which have optional leading numeric marker and length bytes
+                    automata.add(optionalNumericCharPrefix());
```

Comment:
@cbuescher This is a quite clever way to organize `wildcard` query!!! Good job here!!!

For this particular case of `?` case, I think we need to add additional optional automata of `PRERELEASE_SEPARATOR_BYTE` and `NO_PRERELEASE_SEPARATOR_BYTE`.  

Otherwise,  a query `"1.2.3??"` can't match  a version like `"1.2.3+b"` or a version like `"1.2.3-a"`

Reply:
Great catch, I'll add a test for this and will work on it